common/lib/driver-definitions/api-report/driver-definitions.api.md

```diff
@@ -44,6 +44,7 @@ export enum DriverErrorType {
     throttlingError = "throttlingError",
     // (undocumented)
     unsupportedClientProtocolVersion = "unsupportedClientProtocolVersion",
+    usageError = "driverUsageError",
```

Comment:
it would be great to add some pointers here to the code where this will be leveraged

Reply:
@anthony-murphy these errors indicate you've created your document storage manager with a bad Resolved URL:

https://github.com/microsoft/FluidFramework/blob/d21a2ab21eb273198afa08ce83081bebcfe24f6a/packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts#L690-L715

Reply:
but this is not really done by the user, the url resolver parses the url into a resolved url, which then is given to the driver and used to make calls. i think if this is a usageError than every error is a usage error, which isn't very useful

Reply:
i would call this something like, invalidResolvedUrl, with seem general and specific enough

Reply:
i think the original problem was that both the driver and container error enums used the same strings. could we have just changed the string here to be "genericDriverError", rather than deprecating, then introducing numerous breaking changes. it just seems like a ton of churn (7 broken types) for something very minor

Reply:
> then every error is a usage error, which isn't very useful

More useful than "genericError" or "genericDriverError".  At least in my opinion ;)  I think the "genericError" type is too tempting to use, and doesn't mean anything.  If all the errors we are throwing (that aren't network errors) from Driver code are due to misuse of the API or passing in invalid params, then yeah they'll all be usage errors... because they're all usage errors.

Some other specific examples I've identified are in the URL Resolver itself, which today either asserts (wrong) or throws `new Error`.  As for the cases I mentioned above, I have a follow-up PR to move these checks to the constructor - this code will never get hit - and that makes it all the more clear that it's a parameter validation kind of scenario.

Reply:
To me it really boils down to the point of the error type. For other driver error type we use them to do error specific handling with layers like the deltamanager. for me, usage error is fail fast, and tell the consuming developer they did something wrong that they should fix. if it is neither of those cases, what is the point? we don't want to unnecessarily profligate error types that are unactionable by use or end users. so it makes sense we would have a lot of generic errors. these are errors that there is not programmatic solution for, and the stack and message should be used to debug.

Reply:
we should avoid over or premature categorization. we should really only categorize things that are useful to categorize and that we have a specific scenario for, otherwise we just end up with a huge supportability matrix for things we don't really need or have a specific use for

Reply:
looking at usage error thoughout the code base i see some usages that are suspect, but we do have one i think is good:
https://github.com/microsoft/FluidFramework/blob/787ab708edc7ca6dad63c988f880b5c4ada485dc/packages/loader/container-loader/src/container.ts#L797

Reply:
> usage error is fail fast, and tell the consuming developer they did something wrong that they should fix.

Yes... we have cases like that in the Driver layer... so what's your objection to adding this here?  I think each layer needs their own usage error so it's distinguishable (user of Driver v. user of Container).

If you don't want to deprecate genericError, let's talk about that elsewhere.

Reply:
Look at the four occurrences of `throw new Error` in odspDriverUrlResolver for additional cases that are usage errors.

Reply:
@anthony-murphy - I opened https://github.com/microsoft/FluidFramework/issues/8821 about the fact that any change to these enums breaks compatibility.  Let's close on that orthogonally to this discussion of whether `driverUsageError` belongs in our list of valid errorTypes.

Reply:
@scottn12 thanks for your patience here.  I will try to circle back on this specific question of whether usageError fits in here and if not, a different specific course of action.